File: app/src/main/java/com/yuki/yukihub/scanner/GameScanner.java
package com.yuki.yukihub.scanner;

import android.content.Context;
import android.net.Uri;
import android.util.Log;

import androidx.documentfile.provider.DocumentFile;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public class GameScanner {
    private static final String TAG = "GameScanner";
    /** Traverse every descendant directory regardless of depth. */
    public static final int SCAN_ALL_LEVELS = -1;
    /** Traverse descendants until a directory is identified as a game, then stop below it. */
    public static final int SCAN_UNTIL_GAME_MATCH = -2;

    public static List<ScanResult> scan(Context context, Uri rootUri) {
        return scan(context, rootUri, 2);
    }

    public static List<ScanResult> scan(Context context, Uri rootUri, int maxDepth) {
        return scan(context, rootUri, ScanRequest.defaults(maxDepth)).getResults();
    }

    /** Runs a bounded scan and returns both partial results and stop/error information. */
    public static ScanReport scan(Context context, Uri rootUri, ScanRequest request) {
        ScanReport report = new ScanReport();
        Set<String> seenUris = new HashSet<>();
        if (context == null || rootUri == null) {
            report.setStopReason(ScanReport.StopReason.INVALID_ROOT);
            report.addError("扫描目录不可用");
            return report;
        }
        ScanRequest safeRequest = request == null ? ScanRequest.defaults(2) : request;
        int requestedDepth = safeRequest.getMaxDepth();
        boolean scanAllLevels = requestedDepth == SCAN_ALL_LEVELS || requestedDepth == SCAN_UNTIL_GAME_MATCH;
        boolean stopAtGameMatch = requestedDepth == SCAN_UNTIL_GAME_MATCH;
        int depth = scanAllLevels ? Integer.MAX_VALUE : Math.max(1, Math.min(4, requestedDepth));

        DocumentFile root;
        try {
            root = DocumentFile.fromTreeUri(context, rootUri);
        } catch (Throwable t) {
            Log.w(TAG, "fromTreeUri failed uri=" + rootUri, t);
            report.setStopReason(ScanReport.StopReason.INVALID_ROOT);
            report.addError("无法访问扫描目录：" + rootUri);
            return report;
        }
        if (root == null) {
            report.setStopReason(ScanReport.StopReason.INVALID_ROOT);
            report.addError("扫描目录不存在：" + rootUri);
            return report;
        }
        try {
            if (!root.isDirectory()) {
                report.setStopReason(ScanReport.StopReason.INVALID_ROOT);
                report.addError("扫描目标不是目录：" + rootUri);
                return report;
            }
        } catch (Throwable t) {
            Log.w(TAG, "root isDirectory failed uri=" + rootUri, t);
            report.setStopReason(ScanReport.StopReason.INVALID_ROOT);
            report.addError("无法读取扫描目录：" + rootUri);
            return report;
        }
        // A user may select one game directory itself rather than its parent. Probe the
        // root before traversing children so that Kirikiri/ONS/Tyrano/Artemis roots are
        // not skipped merely because they have no game-directory child.
        boolean rootGameMatched = detectGameDirectory(root, report, seenUris, safeRequest);
        if (!(stopAtGameMatch && rootGameMatched)) {
            scanChildren(root, 1, depth, stopAtGameMatch, report, seenUris, safeRequest);
        }
        return report;
    }

    private static boolean detectGameDirectory(DocumentFile dir, ScanReport report, Set<String> seenUris, ScanRequest request) {
        if (dir == null || report == null || !report.tryVisit(request, safeUri(dir))) return false;
        try {
            EngineDetector.Result detected = EngineDetector.detect(dir, 2);
            if (detected == null || detected.confidence <= 0 || !isRootDirectoryEngine(detected.engine)) return false;
            String uri = dir.getUri().toString();
            if (markSeen(seenUris, uri)) {
                report.addResult(new ScanResult(safeName(dir), uri, detected.engine, detected.confidence,
                        detected.launchTarget, "", detected.xp3Candidates));
            }
            return true;
        } catch (Throwable t) {
            Log.w(TAG, "root engine detection failed uri=" + safeUri(dir), t);
            report.addError("识别扫描目录失败：" + safeUri(dir));
            return false;
        }
    }

    /** PSP and Winlator roots are already emitted per entry file, not as a directory entry. */
    private static boolean isRootDirectoryEngine(com.yuki.yukihub.model.EngineType engine) {
        return engine == com.yuki.yukihub.model.EngineType.KIRIKIRI
                || engine == com.yuki.yukihub.model.EngineType.ONS
                || engine == com.yuki.yukihub.model.EngineType.TYRANO
                || engine == com.yuki.yukihub.model.EngineType.ARTEMIS;
    }

    private static void scanChildren(DocumentFile dir, int level, int maxDepth, boolean stopAtGameMatch, ScanReport report, Set<String> seenUris, ScanRequest request) {
        if (dir == null || report == null || report.shouldStop(request)) return;
        DocumentFile[] children;
        try {
            children = dir.listFiles();
        } catch (Throwable t) {
            Log.w(TAG, "listFiles failed uri=" + safeUri(dir), t);
            report.addError("无法读取目录：" + safeUri(dir));
            return;
        }
        if (children == null) return;

        for (DocumentFile child : children) {
            if (!report.tryVisit(request, safeUri(child))) return;
            if (child == null) continue;
            try {
                if (child.isFile()) {
                    String name = safeName(child);
                    String lowerName = name.toLowerCase(Locale.ROOT);
                    // 情况1：单个PSP文件在根目录
                    if (lowerName.endsWith(".iso") || lowerName.endsWith(".cso") || lowerName.endsWith(".chd") ||
                        lowerName.endsWith(".elf") || lowerName.endsWith(".pbp")) {
                        addPspFileResult(report, seenUris, child, name);
                        continue;
                    }
                    // 情况1：单个 Nintendo 3DS 文件在根目录
                    if (isN3dsFile(lowerName)) {
                        addN3dsFileResult(report, seenUris, child, name);
                        continue;
                    }
                    if (lowerName.endsWith(".desktop")) {
                        addDesktopResult(report, seenUris, stripDesktopSuffix(name), child.getUri().toString(), name, "");
                    }
                    continue;
                }
                if (!child.isDirectory()) continue;

                // 识别目录本身的 PSP / desktop 入口；是否继续遍历由扫描模式决定。
                // 全层模式会继续扫描嵌套游戏，命中模式则在识别游戏目录后停止向下。
                boolean pspDirectory = tryAddPspDirectory(child, report, seenUris);
                boolean n3dsDirectory = tryAddN3dsDirectory(child, report, seenUris);
                boolean desktopDirectory = tryAddDesktopDirectory(child, report, seenUris);

                String childName = safeName(child).toLowerCase(Locale.ROOT);
                boolean internalAssetDir = isInternalAssetDir(childName);

                boolean gameDirectoryMatched = pspDirectory || n3dsDirectory || desktopDirectory;
                if (!internalAssetDir && !gameDirectoryMatched) {
                    EngineDetector.Result detected = EngineDetector.detect(child, 2);
                    if (detected != null && detected.confidence > 0) {
                        String uri = child.getUri().toString();
                        if (markSeen(seenUris, uri)) {
                            report.addResult(new ScanResult(safeName(child), uri, detected.engine, detected.confidence,
                                    detected.launchTarget, "", detected.xp3Candidates));
                        }
                        gameDirectoryMatched = true;
                    }
                }

                if (level < maxDepth && !(stopAtGameMatch && gameDirectoryMatched)) {
                    scanChildren(child, level + 1, maxDepth, stopAtGameMatch, report, seenUris, request);
                }
            } catch (Throwable t) {
                Log.w(TAG, "scan child failed uri=" + safeUri(child), t);
                report.addError("扫描项目失败：" + safeUri(child));
            }
        }
    }

    private static boolean tryAddDesktopDirectory(DocumentFile dir, ScanReport report, Set<String> seenUris) {
        if (dir == null || report == null) return false;
        try {
            DocumentFile[] files = dir.listFiles();
            if (files == null || files.length == 0) return false;

            List<DocumentFile> desktops = new ArrayList<>();
            for (DocumentFile f : files) {
                if (f == null || !f.isFile()) continue;
                String name = safeName(f).toLowerCase(Locale.ROOT);
                if (name.endsWith(".desktop")) desktops.add(f);
            }
            if (desktops.isEmpty()) return false;

            String coverUri = "";
            DocumentFile folderCover = findBestImageInDir(dir);
            if (folderCover != null) coverUri = folderCover.getUri().toString();

            if (desktops.size() == 1) {
                // 情况2：文件夹内只有一个 desktop，标题取文件夹名，但入口仍然是 .desktop 文件本身。
                DocumentFile desktop = desktops.get(0);
                return addDesktopResult(report, seenUris, safeName(dir), desktop.getUri().toString(), safeName(desktop), coverUri);
            }

            // 情况3：文件夹里有多个 desktop，按多个单独条目识别。
            boolean added = false;
            for (DocumentFile desktop : desktops) {
                String name = safeName(desktop);
                added |= addDesktopResult(report, seenUris, stripDesktopSuffix(name), desktop.getUri().toString(), name, coverUri);
            }
            return added;
        } catch (Throwable t) {
            Log.w(TAG, "tryAddDesktopDirectory failed uri=" + safeUri(dir), t);
            return false;
        }
    }

    private static boolean addDesktopResult(ScanReport report, Set<String> seenUris, String title, String resultUri, String launchTarget, String coverUri) {
        if (report == null || resultUri == null || !markSeen(seenUris, resultUri)) return false;
        report.addResult(new ScanResult(
                title == null || title.trim().isEmpty() ? "未命名游戏" : title,
                resultUri,
                com.yuki.yukihub.model.EngineType.WINLATOR,
                90,
                launchTarget,
                coverUri
        ));
        return true;
    }

    private static boolean addPspFileResult(ScanReport report, Set<String> seenUris, DocumentFile pspFile, String fileName) {
        if (report == null || pspFile == null) return false;
        String uri = pspFile.getUri().toString();
        if (!markSeen(seenUris, uri)) return false;
        
        // 从文件名中提取游戏标题（去掉扩展名）
        String title = fileName;
        int dotIndex = title.lastIndexOf('.');
        if (dotIndex > 0) {
            title = title.substring(0, dotIndex);
        }
        
        report.addResult(new ScanResult(
                title == null || title.trim().isEmpty() ? "未命名PSP游戏" : title,
                uri,
                com.yuki.yukihub.model.EngineType.PSP,
                95,
                fileName, // launchTarget设置为文件名
                ""
        ));
        return true;
    }

    /**
     * 尝试添加文件夹里的PSP游戏文件
     * 情况2：文件夹里只有1个PSP文件，游戏名取文件夹名，但入口仍然是PSP文件本身
     * 情况3：文件夹里有多个PSP文件，按多个单独条目识别
     */
    private static boolean tryAddPspDirectory(DocumentFile dir, ScanReport report, Set<String> seenUris) {
        if (dir == null || report == null) return false;
        try {
            DocumentFile[] files = dir.listFiles();
            if (files == null || files.length == 0) return false;

            List<DocumentFile> pspFiles = new ArrayList<>();
            for (DocumentFile f : files) {
                if (f == null || !f.isFile()) continue;
                String name = safeName(f).toLowerCase(Locale.ROOT);
                if (name.endsWith(".iso") || name.endsWith(".cso") || name.endsWith(".chd") || 
                    name.endsWith(".elf") || name.endsWith(".pbp")) {
                    pspFiles.add(f);
                }
            }
            if (pspFiles.isEmpty()) return false;

            String coverUri = "";
            DocumentFile folderCover = findBestImageInDir(dir);
            if (folderCover != null) coverUri = folderCover.getUri().toString();

            if (pspFiles.size() == 1) {
                // 情况2：文件夹内只有一个 PSP文件，标题取文件夹名，但入口仍然是 PSP文件本身。
                DocumentFile pspFile = pspFiles.get(0);
                return addPspFileResultWithCover(report, seenUris, safeName(dir), pspFile.getUri().toString(), safeName(pspFile), coverUri);
            }

            // 情况3：文件夹里有多个 PSP文件，按多个单独条目识别。
            boolean added = false;
            for (DocumentFile pspFile : pspFiles) {
                String name = safeName(pspFile);
                String title = name;
                int dotIndex = title.lastIndexOf('.');
                if (dotIndex > 0) {
                    title = title.substring(0, dotIndex);
                }
                added |= addPspFileResultWithCover(report, seenUris, title, pspFile.getUri().toString(), name, coverUri);
            }
            return added;
        } catch (Throwable t) {
            Log.w(TAG, "tryAddPspDirectory failed uri=" + safeUri(dir), t);
            return false;
        }
    }

    private static boolean addPspFileResultWithCover(ScanReport report, Set<String> seenUris, String title, String resultUri, String launchTarget, String coverUri) {
        if (report == null || resultUri == null || !markSeen(seenUris, resultUri)) return false;
        report.addResult(new ScanResult(
                title == null || title.trim().isEmpty() ? "未命名PSP游戏" : title,
                resultUri,
                com.yuki.yukihub.model.EngineType.PSP,
                95,
                launchTarget,
                coverUri
        ));
        return true;
    }

    private static boolean isN3dsFile(String lowerName) {
        if (lowerName == null) return false;
        return lowerName.endsWith(".3ds") || lowerName.endsWith(".cci") || lowerName.endsWith(".zcci") ||
                lowerName.endsWith(".cxi") || lowerName.endsWith(".zcxi") || lowerName.endsWith(".cia") ||
                lowerName.endsWith(".zcia") || lowerName.endsWith(".3dsx") || lowerName.endsWith(".z3dsx");
    }

    private static boolean addN3dsFileResult(ScanReport report, Set<String> seenUris, DocumentFile n3dsFile, String fileName) {
        if (report == null || n3dsFile == null) return false;
        String uri = n3dsFile.getUri().toString();
        if (!markSeen(seenUris, uri)) return false;

        // 从文件名中提取游戏标题（去掉扩展名）
        String title = fileName;
        int dotIndex = title.lastIndexOf('.');
        if (dotIndex > 0) {
            title = title.substring(0, dotIndex);
        }

        report.addResult(new ScanResult(
                title == null || title.trim().isEmpty() ? "未命名3DS游戏" : title,
                uri,
                com.yuki.yukihub.model.EngineType.NINTENDO_3DS,
                95,
                fileName,
                ""
        ));
        return true;
    }

    /**
     * 尝试添加文件夹里的 Nintendo 3DS 游戏文件
     * 情况2：文件夹里只有1个3DS文件，游戏名取文件夹名，但入口仍然是3DS文件本身
     * 情况3：文件夹里有多个3DS文件，按多个单独条目识别
     */
    private static boolean tryAddN3dsDirectory(DocumentFile dir, ScanReport report, Set<String> seenUris) {
        if (dir == null || report == null) return false;
        try {
            DocumentFile[] files = dir.listFiles();
            if (files == null || files.length == 0) return false;

            List<DocumentFile> n3dsFiles = new ArrayList<>();
            for (DocumentFile f : files) {
                if (f == null || !f.isFile()) continue;
                String name = safeName(f).toLowerCase(Locale.ROOT);
                if (isN3dsFile(name)) {
                    n3dsFiles.add(f);
                }
            }
            if (n3dsFiles.isEmpty()) return false;

            String coverUri = "";
            DocumentFile folderCover = findBestImageInDir(dir);
            if (folderCover != null) coverUri = folderCover.getUri().toString();

            if (n3dsFiles.size() == 1) {
                // 情况2：文件夹内只有一个 3DS 文件，标题取文件夹名，但入口仍然是 3DS 文件本身。
                DocumentFile n3dsFile = n3dsFiles.get(0);
                return addN3dsFileResultWithCover(report, seenUris, safeName(dir), n3dsFile.getUri().toString(), safeName(n3dsFile), coverUri);
            }

            // 情况3：文件夹里有多个 3DS 文件，按多个单独条目识别。
            boolean added = false;
            for (DocumentFile n3dsFile : n3dsFiles) {
                String name = safeName(n3dsFile);
                String title = name;
                int dotIndex = title.lastIndexOf('.');
                if (dotIndex > 0) {
                    title = title.substring(0, dotIndex);
                }
                added |= addN3dsFileResultWithCover(report, seenUris, title, n3dsFile.getUri().toString(), name, coverUri);
            }
            return added;
        } catch (Throwable t) {
            Log.w(TAG, "tryAddN3dsDirectory failed uri=" + safeUri(dir), t);
            return false;
        }
    }

    private static boolean addN3dsFileResultWithCover(ScanReport report, Set<String> seenUris, String title, String resultUri, String launchTarget, String coverUri) {
        if (report == null || resultUri == null || !markSeen(seenUris, resultUri)) return false;
        report.addResult(new ScanResult(
                title == null || title.trim().isEmpty() ? "未命名3DS游戏" : title,
                resultUri,
                com.yuki.yukihub.model.EngineType.NINTENDO_3DS,
                95,
                launchTarget,
                coverUri
        ));
        return true;
    }

    private static DocumentFile findBestImageInDir(DocumentFile dir) {
        if (dir == null || !dir.isDirectory()) return null;
        try {
            DocumentFile[] files = dir.listFiles();
            if (files == null) return null;
            DocumentFile best = null;
            int bestScore = Integer.MIN_VALUE;
            for (DocumentFile f : files) {
                if (f == null || !f.isFile()) continue;
                String name = safeName(f);
                if (!isImageFile(name)) continue;
                int score = coverNameScore(name);
                if (best == null || score > bestScore) {
                    best = f;
                    bestScore = score;
                }
            }
            return best;
        } catch (Throwable ignored) {
            return null;
        }
    }

    private static boolean isImageFile(String name) {
        if (name == null) return false;
        String lower = name.toLowerCase(Locale.ROOT);
        return lower.endsWith(".jpg") || lower.endsWith(".jpeg") || lower.endsWith(".png") || lower.endsWith(".webp") || lower.endsWith(".bmp");
    }

    private static int coverNameScore(String name) {
        if (name == null) return 0;
        String lower = name.toLowerCase(Locale.ROOT);
        if (lower.equals("cover.jpg") || lower.equals("cover.png") || lower.equals("cover.webp")) return 100;
        if (lower.equals("folder.jpg") || lower.equals("folder.png") || lower.equals("folder.webp")) return 95;
        if (lower.contains("cover") || lower.contains("folder") || lower.contains("封面")) return 80;
        if (lower.contains("poster") || lower.contains("package") || lower.contains("main")) return 60;
        return 10;
    }

    private static boolean isInternalAssetDir(String name) {
        if (name == null) return false;
        return name.equals("data") || name.equals("tyrano") || name.equals("resources") || name.equals("arc")
                || name.equals("scenario") || name.equals("system") || name.equals("bgimage") || name.equals("fgimage")
                || name.equals("image") || name.equals("sound") || name.equals("bgm") || name.equals("voice") || name.equals("video")
                || name.equals("movie") || name.equals("font") || name.equals("others") || name.equals("app");
    }

    private static String safeName(DocumentFile file) {
        try {
            String name = file == null ? null : file.getName();
            return name == null || name.trim().isEmpty() ? "未命名游戏" : name;
        } catch (Throwable t) {
            Log.w(TAG, "safeName failed uri=" + safeUri(file), t);
            return "未命名游戏";
        }
    }

    private static boolean markSeen(Set<String> seenUris, String uri) {
        if (seenUris == null) return true;
        String key = com.yuki.yukihub.data.GameRepository.normalizeRootUriKey(uri);
        if (key.isEmpty()) return true;
        return seenUris.add(key);
    }

    private static String stripDesktopSuffix(String name) {
        if (name == null) return "未命名游戏";
        return name.toLowerCase(Locale.ROOT).endsWith(".desktop") ? name.substring(0, Math.max(0, name.length() - 8)) : name;
    }

    private static String safeUri(DocumentFile file) {
        try {
            return file == null || file.getUri() == null ? "null" : file.getUri().toString();
        } catch (Throwable ignored) {
            return "unknown";
        }
    }
}

File: app/src/main/java/com/core/scanner/GameScanner.kt
package com.core.scanner

import android.content.Context
import android.net.Uri
import android.os.SystemClock
import android.util.Log
import com.core.data.GameRepository
import com.core.model.EngineType
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.Phaser

/**
 * Bounded, cache-assisted game scanner.
 *
 * Directory metadata is read through [ScanDirectoryReader], which guarantees one directory
 * listing and no per-entry DocumentFile metadata queries. Directory work uses four workers;
 * SAF provider queries are independently limited to two concurrent calls.
 */
object GameScanner {
    private const val TAG = "GameScanner"
    private const val SCAN_WORKER_COUNT = 4

    const val SCAN_ALL_LEVELS = -1
    const val SCAN_UNTIL_GAME_MATCH = -2

    @JvmStatic
    fun scan(context: Context, rootUri: Uri): List<ScanResult> = scan(context, rootUri, 2)

    @JvmStatic
    fun scan(context: Context, rootUri: Uri, maxDepth: Int): List<ScanResult> =
        scan(context, rootUri, ScanRequest.defaults(maxDepth)).results

    @JvmStatic
    fun scan(context: Context, rootUri: Uri, request: ScanRequest): ScanReport {
        val startedAt = SystemClock.elapsedRealtime()
        val report = ScanReport()
        val requestedDepth = request.maxDepth
        val unbounded = requestedDepth == SCAN_ALL_LEVELS || requestedDepth == SCAN_UNTIL_GAME_MATCH
        val maxDepth = if (unbounded) Int.MAX_VALUE else requestedDepth.coerceIn(1, 4)
        val traverseMatchedGames = shouldTraverseMatchedGames(requestedDepth)
        val reader = ScanDirectoryReader(context.applicationContext, rootUri, request)
        val root = runCatching { reader.root() }.getOrNull()
        if (root == null) {
            report.setStopReason(ScanReport.StopReason.INVALID_ROOT)
            report.addError("无法访问扫描目录：$rootUri")
            return report
        }
        Log.i(
            TAG,
            "scan start root=${root.uri} mode=${if (root.localFile != null) "FILE" else "SAF"} " +
                "workers=$SCAN_WORKER_COUNT cache=${request.isCacheEnabled}"
        )

        val cached = if (request.isCacheEnabled) {
            ScanCacheStore.load(context.applicationContext, rootUri.toString(), requestedDepth)
        } else {
            emptyMap()
        }
        val seenUris: MutableSet<String> = ConcurrentHashMap.newKeySet()
        val executor = Executors.newFixedThreadPool(SCAN_WORKER_COUNT)
        val phaser = Phaser(1)
        lateinit var schedule: (ScanNode, Int, Boolean) -> Unit
        schedule = { directory, level, selectedRoot ->
            if (!report.shouldStop(request)) {
                phaser.register()
                executor.execute {
                    try {
                        scanDirectory(
                            directory = directory,
                            level = level,
                            selectedRoot = selectedRoot,
                            maxDepth = maxDepth,
                            traverseMatchedGames = traverseMatchedGames,
                            reader = reader,
                            cached = cached,
                            report = report,
                            seenUris = seenUris,
                            request = request,
                            schedule = schedule
                        )
                    } finally {
                        phaser.arriveAndDeregister()
                    }
                }
            }
        }

        if (report.tryVisit(request, root.uri.toString())) schedule(root, 0, true)
        phaser.arriveAndAwaitAdvance()
        executor.shutdown()
        report.shouldStop(request)
        if (!report.isPartial) {
            ScanCacheStore.save(
                context.applicationContext,
                rootUri.toString(),
                requestedDepth,
                report.results
            )
        }
        Log.i(
            TAG,
            "scan finish root=${root.uri} elapsedMs=${SystemClock.elapsedRealtime() - startedAt} " +
                "visited=${report.visitedNodes} found=${report.results.size} stop=${report.stopReason}"
        )
        return report
    }

    private fun scanDirectory(
        directory: ScanNode,
        level: Int,
        selectedRoot: Boolean,
        maxDepth: Int,
        traverseMatchedGames: Boolean,
        reader: ScanDirectoryReader,
        cached: Map<String, ScanResult>,
        report: ScanReport,
        seenUris: MutableSet<String>,
        request: ScanRequest,
        schedule: (ScanNode, Int, Boolean) -> Unit
    ) {
        if (report.shouldStop(request)) return

        val cachedDirectory = cached[normalized(directory.uri)]
            ?.takeIf { isDirectoryEngine(it.engine) }
        if (cachedDirectory != null) {
            addResult(report, seenUris, cachedDirectory)
            if (!traverseMatchedGames) return
        }

        val children = try {
            reader.listChildren(directory)
        } catch (t: Exception) {
            if (!request.isCancelled && !request.isDeadlineReached) {
                Log.w(TAG, "list directory failed uri=${directory.uri}", t)
                report.addError("无法读取目录：${directory.uri}")
            }
            return
        }
        if (report.shouldStop(request)) return

        val psp = ArrayList<ScanNode>()
        val n3ds = ArrayList<ScanNode>()
        val switch = ArrayList<ScanNode>()
        val desktops = ArrayList<ScanNode>()
        val childDirectories = ArrayList<ScanNode>()
        var bestCover: ScanNode? = null
        var bestCoverScore = Int.MIN_VALUE

        // One classification pass for every direct child.
        for (node in children) {
            if (!report.tryVisit(request, node.uri.toString())) return
            val lower = node.name.lowercase(Locale.ROOT)
            if (node.isDirectory) {
                childDirectories.add(node)
                continue
            }
            when {
                isPspFile(lower) -> psp.add(node)
                isN3dsFile(lower) -> n3ds.add(node)
                isNintendoSwitchFile(lower) -> switch.add(node)
                lower.endsWith(".desktop") -> desktops.add(node)
            }
            if (isImageFile(lower)) {
                val score = coverNameScore(lower)
                if (bestCover == null || score > bestCoverScore) {
                    bestCover = node
                    bestCoverScore = score
                }
            }
        }

        val coverUri = bestCover?.uri?.toString().orEmpty()
        var fileEntryMatched = false
        fileEntryMatched = emitFileGroup(
            report, seenUris, directory, selectedRoot, psp, EngineType.PSP,
            "未命名PSP游戏", coverUri
        ) || fileEntryMatched
        fileEntryMatched = emitFileGroup(
            report, seenUris, directory, selectedRoot, n3ds, EngineType.NINTENDO_3DS,
            "未命名3DS游戏", coverUri
        ) || fileEntryMatched
        fileEntryMatched = emitFileGroup(
            report, seenUris, directory, selectedRoot, switch, EngineType.NINTENDO_SWITCH,
            "未命名Switch游戏", coverUri
        ) || fileEntryMatched
        fileEntryMatched = emitDesktopGroup(
            report, seenUris, directory, selectedRoot, desktops, coverUri
        ) || fileEntryMatched

        // Entry files directly under the user-selected scan root are independent games. They
        // must not turn the scan root itself into a matched game and prune all sibling folders.
        // For a nested folder, retaining the old behavior is intentional: ROM/desktop entries
        // identify that folder as a game container, so its internal asset folders are skipped.
        var gameMatched = shouldPruneAfterFileMatches(selectedRoot, fileEntryMatched)

        val internalAsset = !selectedRoot &&
            isInternalAssetDir(directory.name.lowercase(Locale.ROOT))
        if (!gameMatched && !internalAsset) {
            val detected = NodeEngineDetector.detect(children, 2, reader::listChildren)
            if (detected.confidence > 0 && isDirectoryEngine(detected.engine)) {
                addResult(
                    report,
                    seenUris,
                    ScanResult(
                        title = directory.name.ifBlank { "未命名游戏" },
                        uri = directory.uri.toString(),
                        engine = detected.engine,
                        confidence = detected.confidence,
                        launchTarget = detected.launchTarget,
                        coverUri = coverUri,
                        xp3Candidates = detected.xp3Candidates,
                        rpgMakerSubtype = detected.rpgMakerSubtype,
                        renpySubtype = detected.renpySubtype,
                        godotSubtype = detected.godotSubtype
                    )
                )
                gameMatched = true
            }
        }

        request.markProgress()
        if (level < maxDepth && (!gameMatched || traverseMatchedGames)) {
            for (child in childDirectories) {
                if (report.shouldStop(request)) return
                schedule(child, level + 1, false)
            }
        }
    }

    private fun emitFileGroup(
        report: ScanReport,
        seenUris: MutableSet<String>,
        directory: ScanNode,
        selectedRoot: Boolean,
        files: List<ScanNode>,
        engine: EngineType,
        unnamedTitle: String,
        coverUri: String
    ): Boolean {
        if (files.isEmpty()) return false
        for (file in files) {
            val title = if (files.size == 1 && !selectedRoot) {
                directory.name
            } else {
                stripExtension(file.name)
            }.ifBlank { unnamedTitle }
            addResult(
                report,
                seenUris,
                ScanResult(
                    title = title,
                    uri = file.uri.toString(),
                    engine = engine,
                    confidence = 95,
                    launchTarget = file.name,
                    coverUri = coverUri
                )
            )
        }
        return true
    }

    private fun emitDesktopGroup(
        report: ScanReport,
        seenUris: MutableSet<String>,
        directory: ScanNode,
        selectedRoot: Boolean,
        files: List<ScanNode>,
        coverUri: String
    ): Boolean {
        if (files.isEmpty()) return false
        for (file in files) {
            val title = if (files.size == 1 && !selectedRoot) {
                directory.name
            } else {
                stripExtension(file.name)
            }.ifBlank { "未命名游戏" }
            addResult(
                report,
                seenUris,
                ScanResult(
                    title = title,
                    uri = file.uri.toString(),
                    engine = EngineType.WINLATOR,
                    confidence = 90,
                    launchTarget = file.name,
                    coverUri = coverUri
                )
            )
        }
        return true
    }

    private fun addResult(
        report: ScanReport,
        seenUris: MutableSet<String>,
        result: ScanResult
    ): Boolean {
        val uri = result.uri ?: return false
        val key = GameRepository.normalizeRootUriKey(uri)
        if (key.isNotEmpty() && !seenUris.add(key)) return false
        report.addResult(result)
        return true
    }

    private fun normalized(uri: Uri): String = GameRepository.normalizeRootUriKey(uri.toString())

    private fun isDirectoryEngine(engine: EngineType?): Boolean = when (engine) {
        EngineType.KIRIKIRI,
        EngineType.ONS,
        EngineType.TYRANO,
        EngineType.ARTEMIS,
        EngineType.RPGMAKER,
        EngineType.RENPY,
        EngineType.GODOT -> true
        else -> false
    }

    private fun isPspFile(name: String): Boolean =
        name.endsWith(".iso") || name.endsWith(".cso") || name.endsWith(".chd") ||
            name.endsWith(".elf") || name.endsWith(".pbp")

    private fun isN3dsFile(name: String): Boolean =
        name.endsWith(".3ds") || name.endsWith(".cci") || name.endsWith(".zcci") ||
            name.endsWith(".cxi") || name.endsWith(".zcxi") || name.endsWith(".cia") ||
            name.endsWith(".zcia") || name.endsWith(".3dsx") || name.endsWith(".z3dsx")

    private fun isNintendoSwitchFile(name: String): Boolean =
        name.endsWith(".xci") || name.endsWith(".nsp") ||
            name.endsWith(".nca") || name.endsWith(".nro")

    private fun isImageFile(name: String): Boolean =
        name.endsWith(".jpg") || name.endsWith(".jpeg") || name.endsWith(".png") ||
            name.endsWith(".webp") || name.endsWith(".bmp")

    private fun coverNameScore(name: String): Int = when {
        name == "cover.jpg" || name == "cover.png" || name == "cover.webp" -> 100
        name == "folder.jpg" || name == "folder.png" || name == "folder.webp" -> 95
        name.contains("cover") || name.contains("folder") || name.contains("封面") -> 80
        name.contains("poster") || name.contains("package") || name.contains("main") -> 60
        else -> 10
    }

    private fun stripExtension(name: String): String = name.substringBeforeLast('.', name)

    private fun isInternalAssetDir(name: String): Boolean =
        name == "data" || name == "tyrano" || name == "resources" || name == "arc" ||
            name == "scenario" || name == "system" || name == "bgimage" || name == "fgimage" ||
            name == "image" || name == "sound" || name == "bgm" || name == "voice" ||
            name == "video" || name == "movie" || name == "font" || name == "others" ||
            name == "app" || name == "game" || name == "renpy"

    internal fun shouldTraverseMatchedGames(requestedDepth: Int): Boolean =
        requestedDepth == SCAN_ALL_LEVELS

    internal fun shouldPruneAfterFileMatches(selectedRoot: Boolean, fileEntryMatched: Boolean): Boolean =
        fileEntryMatched && !selectedRoot
}
